src/perf_tools/fiber_trace.cr

```diff
@@ -140,52 +145,62 @@ module PerfTools::FiberTrace
   #
   # NOTE: The main fiber of each thread is not shown.
   def self.pretty_log_fibers(io : IO) : Nil
-    uniqs = spawn_stack
-      .map { |fiber, stack| {fiber.name, stack, yield_stack[fiber]?} }
-      .group_by { |_, s, y| {s, y} }
-      .transform_values(&.map { |fiber, _, _| fiber })
-      .to_a
-      .sort_by! { |(s, y), names| {-names.size, s, y || Array(Void*).new} }
-
-    io.puts "| Count | Fibers | Spawn stack | Yield stack |"
-    io.puts "|------:|:-------|:------------|:------------|"
-    uniqs.each do |(s, y), names|
-      s = Exception::CallStack.new(__callstack: s).printable_backtrace
-      y = y.try { |y| Exception::CallStack.new(__callstack: y).printable_backtrace }
-
-      io << "| "
-      io << names.size
-      io << " | "
-      names.compact.join(io, ' ') { |name| PerfTools.md_code_span(io, name) }
-      io << " | "
-      s.join(io, "<br>") { |frame| PerfTools.md_code_span(io, frame) }
-      io << " | "
-      if y
-        y.join(io, "<br>") { |frame| PerfTools.md_code_span(io, frame) }
-      else
-        io << "*N/A*"
+    lock.synchronize do
+      uniqs = spawn_stack
+        .map { |fiber, stack| {fiber.name, stack, yield_stack[fiber]?} }
+        .group_by { |_, s, y| {s, y} }
+        .transform_values(&.map { |fiber, _, _| fiber })
+        .to_a
+        .sort_by! { |(s, y), names| {-names.size, s, y || Array(Void*).new} }
+
+      io.puts "| Count | Fibers | Spawn stack | Yield stack |"
+      io.puts "|------:|:-------|:------------|:------------|"
+      uniqs.each do |(s, y), names|
+        s = Exception::CallStack.new(__callstack: s).printable_backtrace
+        y = y.try { |y| Exception::CallStack.new(__callstack: y).printable_backtrace }
+
+        io << "| "
+        io << names.size
+        io << " | "
+        names.compact.join(io, ' ') { |name| PerfTools.md_code_span(io, name) }
+        io << " | "
+        s.join(io, "<br>") { |frame| PerfTools.md_code_span(io, frame) }
+        io << " | "
+        if y
+          y.join(io, "<br>") { |frame| PerfTools.md_code_span(io, frame) }
+        else
+          io << "*N/A*"
+        end
+        io << " |\n"
       end
-      io << " |\n"
     end
   end
-end
 
-class Fiber
-  def initialize(@name : String? = nil, &@proc : ->)
-    previous_def(name, &proc)
-
-    stack = Array.new(PerfTools::FiberTrace::STACK_DEPTH + PerfTools::FiberTrace::STACK_SKIP_SPAWN, Pointer(Void).null)
+  # :nodoc:
+  macro track_fiber(action, current_fiber)
+    stack = Array.new(PerfTools::FiberTrace::STACK_DEPTH + PerfTools::FiberTrace::STACK_SKIP_{{action.upcase.id}}, Pointer(Void).null)
```

Comment:
**suggestion:** Use fresh variables in macro generated code. Not strictly necessary here, but *the right thing to do*.

```suggestion
    %stack = Array.new(PerfTools::FiberTrace::STACK_DEPTH + PerfTools::FiberTrace::STACK_SKIP_{{action.upcase.id}}, Pointer(Void).null)
```